### ISSUE TITLE
loosen block/inline structure requirements

### ISSUE THREAD

**Issue by ianstormtaylor** [ianstormtaylor/slate#2653]
#### Do you want to request a _feature_ or report a _bug_?

Idea / improvement. 

#### What's the current behavior?

Right now Slate restricts what document structures it accepts to make working with the documents easier to reason about, and to make the code easier to write. This happens to coincide with 90% of use cases and works well. 

However there are some use cases where the restrictions might be too strict. And this is something that people bring up every once in a while.

There is currently no way to remove these restrictions—because the codebase is designed on the assumptions that they guarantee. Without the schema rules, many of the built-in transforms would fail. 

#### What's the expected behavior?

I’m open to loosening these restrictions if someone wants to take it on as a contribution.  

However, this is not an easy fix, as it requires many of the core transforms to be updated, and would probably require some new helper functions to make things easy to work with. 

Not only that, but many interactions without the restrictions have unknown behaviors that would need to be well researched and spec’d with tests.

I think there are two main restrictions we could try to lift:

A. **Inlines could contain block nodes.** Right now they cannot. But there are use cases where you might want to allow block nodes to be nested inside inline nodes. I am unsure what this would mean in terms of editing experience. (This does not mean that blocks and inlines could be co-mingled.)

B. **Blocks and inlines could be co-mingled as children.** Right now the children have be either one or the other, but this restriction could be removed too. Again, I don’t know what this would mean for editing commands and transforms. 

Either of these is technically possible. If your use case requires them, it will take someone who is willing to put in the effort to figure out the new behaviors, write tests to handle the new permutations of transforms, and ensure all existing transforms support the new use cases. 

It’s not a small task, but I’m open to helping merge if someone wants to take it on.

**Comment by TheSpyder** [comment]
Yes, the HTML you posted is the correct way to render a nested list. And as Ian indicated the model doesn’t allow it.

I wonder what the existing nested list plugins do?

**Comment by remingtonlang** [comment]
@TheSpyder is there a nested list plugin that is actually up to date for 0.5X?

I'm going through and trying to implement nested listing with the new doc structure though I'm by no means an expert. Adding the sub items is fairly straight forward:
**onKeyDown**
`if (event.shiftKey) {
      let aboveType = Editor.above(editor, {})[0]['type'] ? Editor.above(editor, {})[0]['type'] : null
      if (event.key === '}' && aboveType == 'list-item') {
        event.preventDefault()
        const newLine = { type: 'numbered-list', children: [{  children: [{text: ''}], type: 'list-item' }]}
        Transforms.insertNodes(editor, newLine);
      }
    }`

However I'm having issues breaking out of the nested part is tricky, the documentation isnt the best...

**Comment by Entkenntnis** [comment]
You can still do

```html
<ul>
    <li>
        <p>Item</p>
        <ul>
            <li>
                <p>Subitem</p>
            </li>
        </ul>
    </li>
</ul>
```

This would be valid html and fits into slates data model.

You can also decide to create a `list-item-child`element instead of the `p` element and skip it while rendering.
